[CI] MultiNodesStatsTests testMultipleNodes failing

**Build Scans:**
- [elasticsearch-intake #43286 / part2](https://gradle-enterprise.elastic.co/s/dnliy4csvsfie)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:monitoring:test" --tests "org.elasticsearch.xpack.monitoring.MultiNodesStatsTests.testMultipleNodes" -Dtests.seed=7C22ABF7346A6F51 -Dtests.locale=scn-IT -Dtests.timezone=Africa/Mogadishu -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.monitoring.MultiNodesStatsTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testMultipleNodes'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: global checkpoint is not up-to-date, expected: 36 but was: 31
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testMultipleNodes
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

Here is the stack trace:
```
MultiNodesStatsTests > testMultipleNodes FAILED
    java.lang.AssertionError: global checkpoint is not up-to-date, expected: 36 but was: 31
        at __randomizedtesting.SeedInfo.seed([7C22ABF7346A6F51]:0)
        at org.elasticsearch.index.seqno.ReplicationTracker.invariant(ReplicationTracker.java:832)
        at org.elasticsearch.index.seqno.ReplicationTracker.updateLocalCheckpoint(ReplicationTracker.java:1321)
        at org.elasticsearch.index.shard.IndexShard.updateLocalCheckpointForShard(IndexShard.java:3020)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.updateLocalCheckpointForShard(TransportReplicationAction.java:1309)
        at org.elasticsearch.action.support.replication.ReplicationOperation.updateCheckPoints(ReplicationOperation.java:387)
        at org.elasticsearch.action.support.replication.ReplicationOperation$2.onResponse(ReplicationOperation.java:295)
        at org.elasticsearch.action.support.replication.ReplicationOperation$2.onResponse(ReplicationOperation.java:291)
        at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:438)
        at org.elasticsearch.action.support.RetryableAction$RetryingListener.onResponse(RetryableAction.java:150)
        at org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:49)
        at org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleResponse(TransportService.java:1852)
        at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1630)
        at org.elasticsearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:465)
        at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:419)
        at org.elasticsearch.transport.InboundHandler.executeResponseHandler(InboundHandler.java:166)
        at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:134)
        at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102)
        at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:831)
        at org.elasticsearch.transport.InboundPipeline.forwardFragment(InboundPipeline.java:116)
        at org.elasticsearch.transport.InboundPipeline.lambda$doHandleBytes$1(InboundPipeline.java:78)
        at org.elasticsearch.transport.InboundDecoder.finishMessage(InboundDecoder.java:144)
        at org.elasticsearch.transport.InboundDecoder.internalDecode(InboundDecoder.java:129)
        at org.elasticsearch.transport.InboundDecoder.decode(InboundDecoder.java:58)
        at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:79)
        at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:69)
        at org.elasticsearch.transport.netty4.Netty4MessageInboundHandler.channelRead(Netty4MessageInboundHandler.java:54)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at org.elasticsearch.transport.netty4.NettyByteBufSizer.channelRead(NettyByteBufSizer.java:33)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:171)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:697)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:660)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.base/java.lang.Thread.run(Thread.java:1516)
```
So it's actually a rare race condition in ReplicationTracker, not a bug in this test. The bug is harmless because assertions are not on in production, and it recovers automatically. It looks like the problem is that the index was created and then deleted while the replica was being recovered. That caused the thread running updateLocalCheckpoint to get interrupted before it called `pendingInSync.remove(allocationId)` and `updateGlobalCheckpointOnPrimary()`, and for the wait in ReplicationTracker's waitForLocalCheckpointToAdvance() to throw an InterruptedException. So I think we just need to call updateGlobalCheckpointOnPrimary() if the `allocatoinId` is in `pendingSync`.

--- a/docs/changelog/154195.yaml
+++ b/docs/changelog/154195.yaml
@@ -1,0 +1,6 @@
+area: Recovery
+issues:
+ - 153497
+pr: 154195
+summary: Fixing `ReplicationTracker` behavior when a replica recovery is cancelled
+type: bug

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -490,9 +490,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.ccq.MultiClusterSpecStatsIT
   method: test {csv-spec:stats.sumWithOverflowGroupingRow}
   issue: https://github.com/elastic/elasticsearch/issues/153434
-- class: org.elasticsearch.xpack.monitoring.MultiNodesStatsTests
-  method: testMultipleNodes
-  issue: https://github.com/elastic/elasticsearch/issues/153497
 - class: org.elasticsearch.xpack.transform.checkpoint.TransformGetCheckpointIT
   method: testGetCheckpointTimeoutExceeded
   issue: https://github.com/elastic/elasticsearch/issues/150934

--- a/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java
+++ b/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java
@@ -1279,7 +1279,15 @@ public class ReplicationTracker extends AbstractIndexShardComponent implements L
                     }
                 }
             } finally {
-                pendingInSync.remove(allocationId);
+                /*
+                 * If the allocation ID is still in pendingInSync when we get here, then we did not complete normally. We bailed out of the
+                 * thread running waitForLocalCheckpointToAdvance(), likely because the recovery was cancelled and the thread was
+                 * interrupted before updateLocalCheckpoint() completed. So we need to recompute the global checkpoint to avoid problems,
+                 * including an assertion in the invariant() method.
+                 */
+                if (pendingInSync.remove(allocationId)) {
+                    updateGlobalCheckpointOnPrimary();
+                }
             }
         } else {
             cps.inSync = true;

--- a/server/src/test/java/org/elasticsearch/index/seqno/ReplicationTrackerTests.java
+++ b/server/src/test/java/org/elasticsearch/index/seqno/ReplicationTrackerTests.java
@@ -706,6 +706,59 @@ public class ReplicationTrackerTests extends ReplicationTrackerTestCase {
         assertThat(tracker.getGlobalCheckpoint(), equalTo((long) nextActiveLocalCheckpoint));
     }
 
+    /**
+     * Mirrors {@link #testRaceUpdatingGlobalCheckpoint} for the case where the thread marking an allocation ID as in-sync is
+     * interrupted (as happens when a peer recovery is cancelled) while it is still pending in-sync and blocking global checkpoint
+     * advancement. Removing the allocation ID from the pending set on the way out unblocks advancement, so the global checkpoint
+     * must be recomputed; otherwise it is left stale and an assertion in {@link ReplicationTracker} is violated on the next operation
+     * with a "global checkpoint is not up-to-date" assertion error.
+     */
+    public void testGlobalCheckpointUpdateAfterInterruptedMarkAsInSync() throws Exception {
+        final AllocationId active = AllocationId.newInitializing();
+        final AllocationId initializing = AllocationId.newInitializing();
+
+        final long activeLocalCheckpoint = randomIntBetween(1, 1024);
+        final ReplicationTracker tracker = newTracker(active);
+        tracker.updateFromMaster(
+            randomNonNegativeLong(),
+            Collections.singleton(active.getId()),
+            routingTable(Collections.singleton(initializing), active)
+        );
+        tracker.activatePrimaryMode(activeLocalCheckpoint);
+        addPeerRecoveryRetentionLease(tracker, initializing);
+        tracker.initiateTracking(initializing.getId());
+
+        // Mark the initializing shard as in-sync with a local checkpoint below the global checkpoint on a separate thread. This
+        // adds it to the pending-in-sync set and blocks the thread waiting for its local checkpoint to advance.
+        final long initializingLocalCheckpoint = randomLongBetween(NO_OPS_PERFORMED, activeLocalCheckpoint - 1);
+        final AtomicBoolean interrupted = new AtomicBoolean();
+        final Thread markingThread = new Thread(() -> {
+            try {
+                tracker.markAllocationIdAsInSync(initializing.getId(), initializingLocalCheckpoint);
+            } catch (final InterruptedException e) {
+                interrupted.set(true);
+            }
+        });
+        markingThread.start();
+
+        assertBusy(() -> assertTrue(tracker.pendingInSync.contains(initializing.getId())));
+
+        // Advance the active (in-sync) shard's local checkpoint. The global checkpoint cannot advance yet because the initializing
+        // shard is still pending in-sync and blocks advancement.
+        final long nextActiveLocalCheckpoint = randomLongBetween(activeLocalCheckpoint + 1, Integer.MAX_VALUE);
+        tracker.updateLocalCheckpoint(active.getId(), nextActiveLocalCheckpoint);
+        assertThat(tracker.getGlobalCheckpoint(), equalTo(activeLocalCheckpoint));
+
+        // Interrupt the marking thread while it is still pending in-sync, simulating a cancelled peer recovery. Removing the
+        // initializing shard from the pending set unblocks advancement, so the global checkpoint must now reflect the advanced
+        // local checkpoint of the sole remaining in-sync shard.
+        markingThread.interrupt();
+        markingThread.join();
+        assertTrue(interrupted.get());
+
+        assertThat(tracker.getGlobalCheckpoint(), equalTo(nextActiveLocalCheckpoint));
+    }
+
     public void testPrimaryContextHandoff() throws IOException {
         final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings("test", Settings.EMPTY);
         final ShardId shardId = new ShardId("test", "_na_", 0);

